src/Algebra/Graph/AdjacencyMap/Algorithm.hs

```diff
@@ -321,19 +326,98 @@ isAcyclic = isRight . topSort
 -- 'isAcyclic' x     == (scc x == 'gmap' NonEmpty.'NonEmpty.vertex' x)
 -- @
 scc :: Ord a => AdjacencyMap a -> AdjacencyMap (NonEmpty.AdjacencyMap a)
-scc m = gmap (component Map.!) $ removeSelfLoops $ gmap (leader Map.!) m
+scc g = condense g $ execState (gabowSCC g) initialState where
+  initialState = SCC 0 0 [] [] Map.empty Map.empty IntMap.empty [] []
+
+data StateSCC a
+  = SCC { preorder      :: {-# unpack #-} !Int
+        , component     :: {-# unpack #-} !Int
+        , boundaryStack :: [(Int,a)]
+        , pathStack     :: [a]
+        , preorders     :: Map.Map a Int
+        , components    :: Map.Map a Int
+        , innerGraphs   :: IntMap.IntMap (List (AdjacencyMap a))
```

Comment:
As clarified in my comment below, I think we don't need to store a lot of small adjacency maps here and instead should just have something liks `innerGraphs :: IntMap.IntMap (AdjacencyMap a)`. I don't see what `List` brings us here.